### ISSUE TITLE
Add `am/pm` to DateFormat 

### ISSUE THREAD

**Issue by ChristianKurz** [JuliaLang/julia#23870]
currently, DateFormat does not support `am/pm`. As this format is often used, it would be useful to be able to convert it to DateTime.

For Minutes after midday (e.g. 12:0-12:59 AM) and after midnight (0:01-0:59 PM), the am/pm-rule is a bit strange. Support in DateTime could help prevent bugs.

**Comment by KlausC** [comment]
It seems rather unclear to me, what would be the "US-Standard" notation.
See [https://en.wikipedia.org/wiki/12-hour_clock#Confusion_at_noon_and_midnight](url)

**Comment by mikesafar** [comment]
This seems like a big gap here in this library.   In my view, there is no question:  12:00 am is midnight, 12 PM is noon.   It is true that many programmers make deterministic statements that, for example 11:59:59 is "end of day", but this is to avoid overlap of data that may try to capture >= 12:00 AM, with a margin of error of 1 second (which by the way is insufficient in a variety of high-precision applications).   Not having this support means fundamentally more checking in the client code.   

Most systems follow this assumption.    

BTW:   AM/PM are Latin abbreviations:

*AM* = Ante meridiem: Before noon
*PM* = Post meridiem: After noon

But if you make the assumption that all the time "labeled" as 12 is in the same AM/PM designation, then you're good.

**Comment by stevengj** [comment]
In Java, [12pm is noon](https://www.javacodex.com/Date-and-Time/Is-12PM-Noon-Or-Midnight).  In Python, `dateutil.parser` also gives noon—the relevant code was apparently added without much discussion [back in 2003](https://github.com/python/cpython/commit/2bdb61479d235339dc92505d5eccaa6accf4d94c).  For the glibc `strftime` function, 12pm is noon, and similarly [for musl](https://github.com/ifduyue/musl/blob/b4b1e10364c8737a632be61582e05a8d3acf5690/src/time/strftime.c#L123) and [BSD libc](https://github.com/freebsd/freebsd/blob/1d6e4247415d264485ee94b59fdbc12e0c566fd0/lib/libc/stdtime/strftime.c#L297-L299), but the [POSIX standard for `strftime`](http://pubs.opengroup.org/onlinepubs/007908799/xsh/strftime.html) does not seem to specify a behavior.  In Ruby, they have their own `strftime` implementation which [also uses 12pm==noon](https://github.com/ruby/ruby/blob/4444025d16ae1a586eee6a0ac9bdd09e33833f3c/ext/date/date_strftime.c#L347-L350).

In short, while 12am/pm are technically ambiguous, I haven't yet been able to find any widely used library that deviates from 12pm==noon.

**Comment by StefanKarpinski** [comment]
It seems to be very widely used in practice that 12 pm is noon and 12 am is midnight. Indeed, it would be very jarring—discontinuous, technically—otherwise since 12:01 pm is definitely in one minute after noon and 12:01 am is definitely one minute after midnight. It's far less weird to have an am/pm discontinuity between 11:59 pm and 12:00 am and likewise between 11:59 am and 12:00 pm.

**Comment by maxbennedich** [comment]
The above code is incorrect for "12 AM". As is [this code](https://discourse.julialang.org/t/am-pm-in-date-format/17356/3) which is based on it. [This code](https://discourse.julialang.org/t/reading-date-jl/20920/2) suggested on Discourse today is wrong for both 12 AM and 12 PM. This really needs to be supported by the parser since it's too easy for users to mess up if doing oneself. Until then, if anyone else finds this issue, here's a (hopefully) corrected version of the above code:
```julia
using Dates

function parse_datetime_ampm(ds::AbstractString, fmt::DateFormat)
    m = match(r"(.*?)\s*(AM|PM)?$"i, ds)
    d = DateTime(m.captures[1], fmt)
    ampm = uppercase(something(m.captures[2], ""))
    d + Hour(12 * +(ampm == "PM", ampm == "" || hour(d) != 12, -1))
end

dx = [
    "May 1, 2001 12:15am",
    "May 1, 2001 00:15",
    "May 1, 2001 1:15 AM",
    "May 1, 2001 01:15",
    "May 1, 2001 12:15 pm",
    "May 1, 2001 12:15",
    "May 1, 2001 1:15PM",
    "May 1, 2001 13:15"]

println.(parse_datetime_ampm.(dx, dateformat"U d, y H:M"));
```

**Comment by galenlynch** [comment]
I would really appreciate it if this were handled in the `Dates` standard library.

**Comment by stevengj** [comment]
I think this is just waiting for someone to make a PR; the consensus seems to be in favor.

**Comment by stevengj** [comment]
Note that you can use `Libc.strptime` to do this already:
```
julia> Time(Libc.strptime("%I:%M%p", "12:03am"))
00:03:00
```
